service/config.go

```diff
@@ -27,61 +23,16 @@ type Config struct {
 	Extensions extensions.Config `mapstructure:"extensions"`
 
 	// Pipelines are the set of data pipelines configured for the service.
-	Pipelines map[component.ID]*PipelineConfig `mapstructure:"pipelines"`
+	Pipelines pipelines.Config `mapstructure:"pipelines"`
 }
 
 func (cfg *Config) Validate() error {
-	// Must have at least one pipeline.
-	if len(cfg.Pipelines) == 0 {
-		return errMissingServicePipelines
-	}
-
-	// Check that all pipelines have at least one receiver and one exporter, and they reference
-	// only configured components.
-	for pipelineID, pipeline := range cfg.Pipelines {
-		if pipelineID.Type() != component.DataTypeTraces && pipelineID.Type() != component.DataTypeMetrics && pipelineID.Type() != component.DataTypeLogs {
-			return fmt.Errorf("service::pipeline::%s: unknown datatype %q", pipelineID, pipelineID.Type())
-		}
-
-		// Validate pipeline has at least one receiver.
-		if err := pipeline.Validate(); err != nil {
-			return fmt.Errorf("service::pipeline::%s: %w", pipelineID, err)
-		}
+	if err := cfg.Pipelines.Validate(); err != nil {
+		return fmt.Errorf("service::pipelines config validation failed: %w", err)
 	}
 
 	if err := cfg.Telemetry.Validate(); err != nil {
-		fmt.Printf("service::telemetry config validation failed, %v\n", err)
-	}
-
-	return nil
-}
-
-// PipelineConfig defines the configuration of a Pipeline.
-type PipelineConfig struct {
-	Receivers  []component.ID `mapstructure:"receivers"`
-	Processors []component.ID `mapstructure:"processors"`
-	Exporters  []component.ID `mapstructure:"exporters"`
-}
-
-func (cfg *PipelineConfig) Validate() error {
-	// Validate pipeline has at least one receiver.
-	if len(cfg.Receivers) == 0 {
-		return errMissingServicePipelineReceivers
-	}
-
-	// Validate pipeline has at least one exporter.
-	if len(cfg.Exporters) == 0 {
-		return errMissingServicePipelineExporters
-	}
-
-	// Validate no processors are duplicated within a pipeline.
-	procSet := make(map[component.ID]struct{}, len(cfg.Processors))
-	for _, ref := range cfg.Processors {
-		// Ensure no processors are duplicated within the pipeline
-		if _, exists := procSet[ref]; exists {
-			return fmt.Errorf("references processor %q multiple times", ref)
-		}
-		procSet[ref] = struct{}{}
+		fmt.Printf("service::telemetry config validation failed: %v\n", err)
```

Comment:
this should return fmt.Errorf()

Reply:
I want to do it in a separate PR, since it changes previous behavior.